# Assignment 7 (Week 7)

>**Note**: 

```
- Late submissions are penalized.
- Only GitHub submissions are acceptable.
```

## Name: Gafar Babatunde BAMIGBADE
<br>

## Please show and display ALL your calculations and results.
> Remember to read the **`instructions`** carefully.

INTRODUCTION

I am conducting a project in data science using Jupyter notebook to explore my knowledge and understanding in 
Pandas, Analytical reasoning,Exploratory Data Analysis, Visualisations using Matplotlib / Seaborn and Machine learning


IDEAS OF THE PROJECT

I am will be exploring some analysis on the salary scale of a particular company to predict if any employee is making over 50K per yera. This analysis will require me to demonstrate some skills including: Feature engineering, Feature selection, Model building, Model evaluation and Code style and code quality

SOURCE OF DATA

The dataset was provided by the tutor via this link (https://drive.google.com/file/d/1_c3KA14xQC02K0QZ4cpi1emjdz0rqHzb/view?usp=share_link). There were no missing values in the data.

LOADING OF NECESSARY LIBARIES

All required libaries and standard imports are done below

In [1]:
# Built-in library
import itertools

# Standard imports
import numpy as np
import pandas as pd
import matplotlib as plt

In [2]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder, OneHotEncoder
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score

In [3]:
# Import necessary libraries
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import (
    accuracy_score,
    precision_score,
    recall_score,
    f1_score,
    roc_auc_score,
)

In [4]:
# pandas settings
pd.options.display.max_rows = 1_000
pd.options.display.max_columns = 1_000
pd.options.display.max_colwidth = 1_000

In [5]:
pip install lab_black

Note: you may need to restart the kernel to use updated packages.


ERROR: Could not find a version that satisfies the requirement lab_black (from versions: none)
ERROR: No matching distribution found for lab_black


In [6]:
# Black code formatter (Optional)
%load_ext nb_black

<IPython.core.display.Javascript object>

> The data can be found [here](https://drive.google.com/file/d/1_c3KA14xQC02K0QZ4cpi1emjdz0rqHzb/view?usp=share_link).

### Data Dictionary

```
- Age: continuous.

- Workclass: Private, Self-emp-not-inc, Self-emp-inc, Federal-gov, Local-gov, State-gov, Without-pay, Never-worked.

- Final_weight: continuous.

- Education: Bachelors, Some-college, 11th, HS-grad, Prof-school, Assoc-acdm, Assoc-voc, 9th, 7th-8th, 12th, - Masters, 1st-4th, 10th, Doctorate, 5th-6th, Preschool.

- Education_num: continuous.

- Marital_status: Married-civ-spouse, Divorced, Never-married, Separated, Widowed, Married-spouse-absent, - Married-AF-spouse.

- Occupation: Tech-support, Craft-repair, Other-service, Sales, Exec-managerial, Prof-specialty, Handlers-cleaners, Machine-op-inspct, Adm-clerical, Farming-fishing, Transport-moving, Priv-house-serv, Protective-serv, Armed-Forces.

- Relationship: Wife, Own-child, Husband, Not-in-family, Other-relative, Unmarried.

- Race: White, Asian-Pac-Islander, Amer-Indian-Eskimo, Other, Black.

- Genger: Female, Male.

- Capital_gain: continuous.

- Capital_loss: continuous.

- Hours_per_week: continuous.

- Country: United-States, Cambodia, England, Puerto-Rico, Canada, Germany, Outlying-US(Guam-USVI-etc), India, Japan, Greece, South, China, Cuba, Iran, Honduras, Philippines, Italy, Poland, Jamaica, Vietnam, Mexico, Portugal, Ireland, France, Dominican-Republic, Laos, Ecuador, Taiwan, Haiti, Columbia, Hungary, Guatemala, Nicaragua, Scotland, Thailand, Yugoslavia, El-Salvador, Trinadad&Tobago, Peru, Hong, Holand-Netherlands.

- Salary: 
```

### Objective

```
Predict whether a person makes over 50K a year.

```



###  Qs 1. Build a machine learning model that predicts the salary.

###  Qs 2. Evaluate the performance of your model using at least three (3) performance metrics.

<hr>

## Note: 

- The assignment **should** be submitted through a `public` GitHub repository.

DATA PROCESSING AND VISULAIZATION

I have imported the data so as to see how the data look like and carry out further analysis.


In [7]:
# Data importation
data = pd.read_csv("salary.csv")
data.head()

Age          Workclass  Final_weight   Education  Education_num  \
0   50   Self-emp-not-inc         83311   Bachelors             13   
1   38            Private        215646     HS-grad              9   
2   53            Private        234721        11th              7   
3   28            Private        338409   Bachelors             13   
4   37            Private        284582     Masters             14   

        Marital_status          Occupation    Relationship    Race      Sex  \
0   Married-civ-spouse     Exec-managerial         Husband   White     Male   
1             Divorced   Handlers-cleaners   Not-in-family   White     Male   
2   Married-civ-spouse   Handlers-cleaners         Husband   Black     Male   
3   Married-civ-spouse      Prof-specialty            Wife   Black   Female   
4   Married-civ-spouse     Exec-managerial            Wife   White   Female   

   Capital_gain  Capital_loss  Hours_per_week         Country  Salary  
0             0             0              13   United-States   <=50K  
1             0             0              40   United-States   <=50K  
2             0             0              40   United-States   <=50K  
3             0             0              40            Cuba   <=50K  
4             0             0              40   United-States   <=50K

<IPython.core.display.Javascript object>

In [8]:
# Confirmation of data shape
data.shape

(32560, 15)

<IPython.core.display.Javascript object>

The data shape revealed a 32,560 rows and 15 columns table information

In [9]:
# Descriptive statistics of numerical variables in the data
data.describe()

Age  Final_weight  Education_num  Capital_gain  Capital_loss  \
count  32560.000000  3.256000e+04   32560.000000  32560.000000  32560.000000   
mean      38.581634  1.897818e+05      10.080590   1077.615172     87.306511   
std       13.640642  1.055498e+05       2.572709   7385.402999    402.966116   
min       17.000000  1.228500e+04       1.000000      0.000000      0.000000   
25%       28.000000  1.178315e+05       9.000000      0.000000      0.000000   
50%       37.000000  1.783630e+05      10.000000      0.000000      0.000000   
75%       48.000000  2.370545e+05      12.000000      0.000000      0.000000   
max       90.000000  1.484705e+06      16.000000  99999.000000   4356.000000   

       Hours_per_week  
count    32560.000000  
mean        40.437469  
std         12.347618  
min          1.000000  
25%         40.000000  
50%         40.000000  
75%         45.000000  
max         99.000000

<IPython.core.display.Javascript object>

The descriptive statistics showed the count, mean, standard deviation, minimum and maximum values of only the numerical variables in the dataset.

Check NaN values

In [10]:
# Checking for the possibilities of missing values
data.isnull().sum()

Age               0
Workclass         0
Final_weight      0
Education         0
Education_num     0
Marital_status    0
Occupation        0
Relationship      0
Race              0
Sex               0
Capital_gain      0
Capital_loss      0
Hours_per_week    0
Country           0
Salary            0
dtype: int64

<IPython.core.display.Javascript object>

From the above output, it was observed that there was no missing values

In [11]:
# Salary column value counts
data["Salary"].value_counts(normalize=True).mul(100)

 <=50K    75.918305
 >50K     24.081695
Name: Salary, dtype: float64

<IPython.core.display.Javascript object>

Since, I am predicting the number of employees making above 50k, the salary column is needed to be counted. From the output, it was observed that a vlaue count of 75.918305 was obtained for salary less than 50K while 24.081695 was obtained for salary greater than 50K

Feature Engineering

I have transformed the numerical variable (salary) and encode categorical variables to display my knowledge of feature engoineering. 



In [12]:
# Salary column encoding
le = LabelEncoder()
data["Salary"] = le.fit_transform(data["Salary"])

<IPython.core.display.Javascript object>

In [13]:
# Storage of encoder for future use
encoder_dict = {}
encoder_dict["Salary"] = le

<IPython.core.display.Javascript object>

In [14]:
# Removing unrelated columns
data.drop(["Education", "Relationship", "Country"], axis=1, inplace=True)

<IPython.core.display.Javascript object>

In [15]:
# Encoding categorical variables
cat_cols = ["Workclass", "Marital_status", "Occupation", "Race", "Sex"]
for col in cat_cols:
    le = LabelEncoder()
    data[col] = le.fit_transform(data[col])
    encoder_dict[col] = le

<IPython.core.display.Javascript object>

Spliting dataset into training and and testing set

In [16]:
# Separating data set in to training and testing set
X = data.drop("Salary", axis=1)
y = data["Salary"]
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

<IPython.core.display.Javascript object>

The building of a machine learning model to predicts the salary using RandomForecastClassifier()

In [17]:
# Selection of  a machine learning algorithm and fitting of the model for training data set
model = RandomForestClassifier()
model.fit(X_train, y_train)

RandomForestClassifier()

<IPython.core.display.Javascript object>

In [18]:
# Using the trained data set model to make predictions for the testing data set
y_pred = model.predict(X_test)

<IPython.core.display.Javascript object>

In [19]:
# Using encoder to convert numeric labels back to non-numeric labels
y_test = encoder_dict["Salary"].inverse_transform(y_test)
y_pred = encoder_dict["Salary"].inverse_transform(y_pred)

<IPython.core.display.Javascript object>

Model performance evaluation with at least three performance metrics

The model performance will be evaluated with:
1. Accuracy score
2. Recall score
3. F1 score
4. Precision score

In [20]:
# Model performance evaluation with accuracy score
acc = accuracy_score(y_test, y_pred)
acc

0.8524262899262899

<IPython.core.display.Javascript object>

In [21]:
# Model performance evaluation with recall score
rec = recall_score(y_test, y_pred, pos_label=" >50K")
rec

0.624375

<IPython.core.display.Javascript object>

In [22]:
# Model performance evaluation with  F1 score
f1 = f1_score(y_test, y_pred, pos_label=" >50K")
f1

0.6752281176072997

<IPython.core.display.Javascript object>

In [23]:
#  Model performance evaluation with precision score
prec = precision_score(y_test, y_pred, pos_label=" >50K")
prec

0.7350993377483444

<IPython.core.display.Javascript object>

Conclusion

Conclusively, the model performance results indicated that accuracy score performance metric is the best metric followed by precision score and F1 score. Interstingly, recall score showed the least performance metric. 
